lib/qos/qos.c

```diff
@@ -0,0 +1,335 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <sys/slist.h>
+
+#include "qos.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(qos, CONFIG_QOS_LOG_LEVEL);
+
+#define INVALID -1
+
+/* Lookup table for QoS backoff messages. */
+uint32_t backoff_delay_lookup_sec[] = {
+	16, 32, 64
+};
+
+/* Structure used to keep track of pending messages. Used in combination with a linked list. */
+struct qos_metadata {
+	/* Mandatory variable used to construct a linked list. */
+	sys_snode_t header;
+
+	/* Message associated with each node in the linked list. */
+	struct qos_data message;
+};
+
+/* Structure containing internal variables in the library.  */
+static struct ctx {
+	/* Library event handler. Used in callbacks to the caller. */
+	qos_evt_handler_t app_evt_handler;
+
+	/* Number of times the message retransmission has been invoked by the library.
+	 * Used to index backoff delay.
+	 */
+	uint8_t retry_count;
```

Comment:
I'm a bit confused here, do we keep a single retry counter for all pending messages? Same goes with the timeout,  shouldn't those values be tracked individually for each pending message?

Reply:
Each individual message structure has a count `notified_count` that is increased every time that particular message is notified by the library.

The library uses a single timer that stays active as long as there are pending messages in the internal list. Each time the timer is triggered the `retry_count` will increase, increasing the backoff. When all pending messages are removed, the timer is reset.

The idea is that all messages should be sent at the same time, within the same RRC connected window. Having separate timers for each message sounds too complex and would be difficult to manage.

When the count expires the application is notified with the QOS_EVT_RETRY_COUNT_EXPIRED event where it can decide if it wants to remove all pending messages or simply reset the internal timer.

I hope this makes sense.

Reply:
Yes, handling timeouts separately for each message would definitely be more complicated. Resending messages at the same time is a good idea at least in theory, but I think it has some downsides as well. I think the biggest problem with the current implementation is (if I understand correctly), that when there are multiple messages, when the retry count is reached, some of the messages might have been sent only once. A message might actually have just been sent, when the retry count is reached immediately and `QOS_EVT_RETRY_COUNT_EXPIRED` is sent to the application.

Reply:
Yes. You are right. That struck me as well... Ill take a closer look and see If I can revise this logic a bit.

Reply:
This can be solved a number of ways. Although, none seem ideal if wanting to keep sending all messages on the same timeout. Some suggestions:

1. Reset the `retry_count` whenever a message is added and the timer is running. This would ensure that every message is resent at least resent `ARRAY_SIZE(backoff_delay_lookup_sec) - 1` times. This means that as long as messages keep getting added to the library, the lowest entry in the `backoff_delay_lookup_sec` table is used. So messages will have to be removed accordingly, based on the `notified_count` parameter perhaps.

2. After every timeout the library could iterate through all messages in the pending list and index the `backoff_delay_lookup_sec` table with the lowest `notified_count` in the list, when choosing the new timeout.

3. 1. Define a larger `backoff_delay_lookup_sec` table.
    2. Define a `resend_max` variable.
    3. When a message is added, the `resend_max` value is extended with with the current value of the `retry_count`.
    4. If a message has been notified x number of times over a configurable threshold, it is removed and the
       `QOS_EVT_MESSAGE_REMOVED_FROM_LIST` is notified with the corresponding message.
    
    If the size of the table is exceeded or `resend_max` equals `retry_count`, the QOS_EVT_RETRY_COUNT_EXPIRED` is notified.

Reply:
What about always using a 16 second timer and whenever that timer expires, checking if each message needs to be retransmitted or if it should be removed from the list? Each message would need to have a timestamp for the last transmission and retry count which would be used to determine the time when the next retransmission should happen.

If multiple messages are sent during a short period of time, the timer would be started when the first message is sent. When the timer expires 16 seconds later, the following messages would have been transmitted a bit less than 16 seconds ago. Because of this it would probably be good to have some tolerance for the time, for example 1 second, so that all messages going to timeout in the next second would be retransmitted at the same time.

In the worst case the first retransmission would happen after 31 (16 + 15) seconds instead of 16 seconds, I think, but some variation is unavoidable if we want to combine the retransmissions.

Reply:
That definitely works and will not require too complex changes. Off course, we loose the beauty of an exponential back-off,  but I don't think that's a must. Ill look into it, thanks.

Reply:
I updated implementation to use a more simple approach. The user can configure a set backoff `CONFIG_QOS_MESSAGE_NOTIFY_TIMEOUT_SECONDS` and a number of times that a message will be notified before it is removed from the library, `CONFIG_QOS_MESSAGE_NOTIFIED_COUNT_MAX`.

Whenever the timeout triggers, the notified count of each pending message is checked against `CONFIG_QOS_MESSAGE_NOTIFIED_COUNT_MAX` and if the count exceeds this value its removed.

The timer is started on the first messages that's added to an empty list and stopped when the last message is removed and the list is empty.

Reply:
Thanks, the new implementation looks good to me. Exponential back-off could have been still used if it would have been message dependent, but that would have made the implementation much more complex.